fs/march/march.go

```diff
@@ -76,13 +77,16 @@ type listDirFn func(dir string) (entries fs.DirEntries, err error)
 
 // makeListDir makes constructs a listing function for the given fs
 // and includeAll flags for marching through the file system.
-func (m *March) makeListDir(ctx context.Context, f fs.Fs, includeAll bool) listDirFn {
+// Note: this will optionally flag filter-aware backends!
+func (m *March) makeListDir(ctx context.Context, f fs.Fs, includeAll, useFilter bool) listDirFn {
```

Comment:
I think `useFilter` should always be `!includeAll`...

Does that seem correct to you?

Reply:
I'm afraid not. `includeAll` is common between march sides and should be passed downstairs unchanged. `useFilter` is `true` only on the source side. Do I miss something?

Reply:
You, the call stack in `march` is so complex that I decided to treat such notions as "march side" and "include all" separately.
The "march side" (aka `useFilter` which is true on the source side)  is passed down the stack in context.
The `includeAll` term is passed as an argument like before. This has an advantage that I don't need to look in every place in the code where it's used.
Probably this can be rewritten differently. See my last commit

Reply:
Note:

I didn't introduce the new context flag in every public march/walk function. Instead, I added them selectively in the functions called from `sync.Copy/Paste`, from `operations.List` and in `Walk`  which is called from some of them. I didn't touch e.g. `list.NewDirList` but modified march initialization instead.

I tagged involved functions by a comment line containing the word `filter-aware` so to simplify analysis a little.

Reply:
To me the call stack in `march` was so complex that I decided to treat such notions as "march side" and "include all" separately:
- passed the "march side" (aka `useFilter` which is true on the source side)  down the stack in context.
- passed `includeAll` term is passed as an argument like before. This has an advantage that I don't need to look in every place in the code where it's used.

Now rewritten to your review. After that I had to modify the unit test (I made it in a sidecar commit) because now query constraining can happen even if drive remote appears on the right side (it's now controlled purely by the march's src/dst includeAll flags).

PTAL!